File: dev-docs/snippets/model-surface/paginate.ts
import { Task } from "../../../examples/_harness/generated/ts/Task.generated";

// Sort and paginate with a cursor. `query()` returns a PaginatedResult; carry
// `.nextCursor` forward via `uniqueStartKey`.
export async function paginate() {
  // #region example
  const page1 = await Task.query(
    { completed: false },
    { limit: 20, sort: { priority: -1 } },
  );

  if (page1.nextCursor) {
    const page2 = await Task.query(
      { completed: false },
      { limit: 20, sort: { priority: -1 }, uniqueStartKey: page1.nextCursor },
    );
    return page2.data;
  }
  // #endregion example
  return page1.data;
}

File: dev-docs/snippets/model-surface/paginate.swift
import JsBaoClient

// `Task.queryPaged` returns a page of rows plus opaque cursors. Pass the
// previous `nextCursor` back via `options.cursor` to walk through a large
// result set page by page — mirroring JS's `query()` `{ data, nextCursor,
// hasMore }` shape.
func paginate() throws {
  // #region example
  var cursor: String? = nil
  repeat {
    let page = try Task.queryPaged(
      ["completed": false],
      options: QueryOptions(sortOrder: [("priority", -1)], limit: 20, cursor: cursor)
    )
    process(page.data)
    cursor = page.hasMore ? page.nextCursor : nil
  } while cursor != nil
  // #endregion example
}

func process(_ rows: [Task]) {}
